packages/math-input/src/components/input/mathquill-helpers.ts

```diff
@@ -0,0 +1,271 @@
+    return mathField.__controller.cursor;
+}
+
+export function isFraction(node): boolean {
+    return node.jQ && node.jQ.hasClass("mq-fraction");
+}
+
+export function isNumerator(node): boolean {
+    return node.jQ && node.jQ.hasClass("mq-numerator");
+}
+
+export function isDenominator(node): boolean {
+    return node.jQ && node.jQ.hasClass("mq-denominator");
+}
+
+export function isSubScript(node): boolean {
+    // NOTE(charlie): MyScript has a structure whereby its superscripts seem
+    // to be represented as a parent node with 'mq-sup-only' containing a
+    // single child with 'mq-sup'.
+    return (
+        node.jQ &&
+        (node.jQ.hasClass("mq-sub-only") || node.jQ.hasClass("mq-sub"))
+    );
+}
+
+export function isSuperScript(node): boolean {
+    // NOTE(charlie): MyScript has a structure whereby its superscripts seem
+    // to be represented as a parent node with 'mq-sup-only' containing a
+    // single child with 'mq-sup'.
+    return (
+        node.jQ &&
+        (node.jQ.hasClass("mq-sup-only") || node.jQ.hasClass("mq-sup"))
+    );
+}
+
+export function isParens(node): boolean {
+    return node && node.ctrlSeq === "\\left(";
+}
+
+export function isLeaf(node): boolean {
+    return node && node.ctrlSeq && ValidLeaves.includes(node.ctrlSeq.trim());
+}
+
+export function isSquareRoot(node): boolean {
+    return (
+        node.blocks &&
+        node.blocks[0].jQ &&
+        node.blocks[0].jQ.hasClass("mq-sqrt-stem")
+    );
+}
+
+export function isNthRoot(node): boolean {
+    return (
+        node.blocks &&
+        node.blocks[0].jQ &&
+        node.blocks[0].jQ.hasClass("mq-nthroot")
+    );
+}
+
+export function isNthRootIndex(node): boolean {
+    return node.jQ && node.jQ.hasClass("mq-nthroot");
+}
+
+export function isInsideLogIndex(cursor: MathFieldCursor): boolean {
+    const grandparent = cursor.parent.parent;
+
+    if (grandparent && grandparent.jQ.hasClass("mq-supsub")) {
+        const command = maybeFindCommandBeforeParens(grandparent);
+
+        if (command && command.name === "\\log") {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+export function isInsideEmptyNode(cursor: MathFieldCursor): boolean {
+    return (
+        cursor[MQ.L] === MathFieldActionType.MQ_END &&
+        cursor[MQ.R] === MathFieldActionType.MQ_END
+    );
+}
+
+export function selectNode(node, cursor) {
+    cursor.insLeftOf(node);
+    cursor.startSelection();
+    cursor.insRightOf(node);
+    cursor.select();
+    cursor.endSelection();
+}
+
+/**
+ * Return the start node, end node, and full name of the command of which
+ * the initial node is a part, or `null` if the node is not part of a
+ * command.
+ *
+ * @param {node} initialNode - the node to included as part of the command
+ * @returns {null|object} - `null` or an object containing the start node
+ *                          (`startNode`), end node (`endNode`), and full
+ *                          name (`name`) of the command
```

Comment:
Not necessary for this PR, but these feel like they'd be great to move to Typescript types instead of being encoded in JSDoc comments. (we could keep the description of the parameter/return, but the type info should be in true type.

Reply:
Yeah, I agree. I don't know where I'd start with that; I don't know what MQ considers a "node" and that feels like something internal to MQ. If it was a quick fix I'd offer to do it now, but not sure this is the right moment.

Reply:
Well, I think the `initialNode` is somewhat obscure, we could type that as `any` for now. But the return value seems well-defined in the comments and code.